src/Compilers/CSharp/Portable/Binder/Binder_Deconstruct.cs

```diff
@@ -466,34 +467,90 @@ private static TypeSymbol MakeMergedTupleType(ArrayBuilder<DeconstructionVariabl
 
         private BoundTupleLiteral DeconstructionVariablesAsTuple(CSharpSyntaxNode syntax, ArrayBuilder<DeconstructionVariable> variables, DiagnosticBag diagnostics, bool hasErrors)
         {
+            bool inferNames = this.Compilation.LanguageVersion >= LanguageVersion.CSharp7_1;
```

Comment:
Consider extracting an `InferTupleElementNames()` helper method since this same check occurs in multiple places.